source/common/router/scoped_config_impl.h

```diff
@@ -1,17 +1,132 @@
 #pragma once
 
+#include <typeinfo>
+
 #include "envoy/api/v2/srds.pb.h"
 #include "envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.pb.h"
 #include "envoy/router/router.h"
 #include "envoy/router/scopes.h"
 #include "envoy/thread_local/thread_local.h"
 
+#include "common/protobuf/utility.h"
 #include "common/router/config_impl.h"
 #include "common/router/scoped_config_manager.h"
 
 namespace Envoy {
 namespace Router {
 
+using envoy::config::filter::network::http_connection_manager::v2::ScopedRoutes;
+
+/**
+ * Scope key fragment base class.
+ */
+class ScopeKeyFragmentBase {
+public:
+  bool operator!=(const ScopeKeyFragmentBase& other) const { return !(*this == other); }
+
+  bool operator==(const ScopeKeyFragmentBase& other) const {
```

Comment:
Question; how important is equality outside of tests? If it's for test only, I would encourage moving this to test utils.

Reply:
This will be used in ActiveStream::decodeHeaders (headers, endstream) before route selection.